engine/runner/src/main/java/org/enso/runner/Utils.java

```diff
@@ -66,4 +73,86 @@ static scala.Tuple3<Boolean, File, String> findFileAndProject(String path, Strin
     }
     return Tuple3.apply(projectMode, canonicalFile, projectRoot);
   }
+
+  static <S> void printStackTrace(
+      Iterable<S> stack,
+      boolean syntaxError,
+      String msg,
+      File relativeTo,
+      Consumer<String> print,
+      Function<S, String> fnLangId,
+      Function<S, String> fnRootName,
+      Function<S, SourceSection> fnSourceSection) {
+
+    var fullStackReversed = new LinkedList<S>();
+    for (var e : stack) {
+      fullStackReversed.addFirst(e);
+    }
+
+    var dropInitJava =
+        new ArrayList<S>(
+            fullStackReversed.stream()
+                .dropWhile(f -> !LanguageInfo.ID.equals(fnLangId.apply(f)))
+                .toList());
+    Collections.reverse(dropInitJava);
```

Comment:
This is the changed line. The rest is just a refactoring to make things testable.